runtime/j9vm/java11vmi.c

```diff
@@ -1634,58 +1634,57 @@ JVM_SetBootLoaderUnnamedModule(JNIEnv *env, jobject module)
+		/* fetch module object reference after GC point */
+		j9object_t modObj = J9_JNI_UNWRAP_REFERENCE(module);
 		J9Class *instanceClazz = J9OBJECT_CLAZZ(currentThread, modObj);
-		if (NULL == currentThread->currentException) {
-			J9Class *moduleClass = vmFuncs->internalFindKnownClass(currentThread,
-					J9VMCONSTANTPOOL_JAVALANGMODULE,
-					J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
-			if (!isModuleUnnamed(currentThread, modObj)) {
-				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "named module was supplied");
-			} else if (!isSameOrSuperClassOf(moduleClass, instanceClazz)) {
-				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module supplied is not same or sub class of java/lang/Module");
-			} else if (instanceClazz->classLoader != systemClassLoader) {
-				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module was not loaded by the bootclassloader");
-			} else {
+		if (!isModuleUnnamed(currentThread, modObj)) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "named module was supplied");
+		} else if (!isSameOrSuperClassOf(moduleClass, instanceClazz)) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module supplied is not same or sub class of java/lang/Module");
+		} else if (instanceClazz->classLoader != systemClassLoader) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, "module was not loaded by the bootclassloader");
+		} else {
 #if JAVA_SPEC_VERSION >= 21
-				J9Module *unnamedModuleForSystemLoader = vm->unnamedModuleForSystemLoader;
-				/* can't set bootclassloader.unnamedModule since bootclassloader hasn't finished the initialization yet */
-				if (NULL == unnamedModuleForSystemLoader) {
-					vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "unnamedModuleForSystemLoader was not initialized");
-				} else if (NULL != unnamedModuleForSystemLoader->moduleObject) {
-					vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "module is already set in the unnamedModuleForSystemLoader");
+			J9Module *unnamedModuleForSystemLoader = vm->unnamedModuleForSystemLoader;
+			/* can't set bootclassloader.unnamedModule since bootclassloader hasn't finished the initialization yet */
+			if (NULL == unnamedModuleForSystemLoader) {
+				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "unnamedModuleForSystemLoader was not initialized");
+			} else if (NULL != unnamedModuleForSystemLoader->moduleObject) {
+				vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGINTERNALERROR, "module is already set in the unnamedModuleForSystemLoader");
+			} else {
+				J9Module *j9mod = NULL;
+				if (IS_RESTORE_RUN(vm)) {
```

Comment:
unrelated but `IS_RESTORE_RUN` needs to be wrapped in `defined(J9VM_OPT_SNAPSHOTS)` @babsingh

Reply:
When the feature is disabled, `IS_RESTORE_RUN` is always `FALSE`. The previous assumption was that the compiler will optimize this and evaluate it as dead code, removing it. However, we've observed perf issues on zLinux where the compiler fails to optimize correctly. Just to be safe, it should be #ifdef'ed.

Reply:
> Just to be safe, it should be #ifdef'ed.

I can open a separated PR for that change.

Reply:
Created https://github.com/eclipse-openj9/openj9/pull/21678
@babsingh please review.